File: core/src/com/alta_v2/rendering/common/component/animation/FadeAnimation.java
package com.alta_v2.rendering.common.component.animation;

import com.google.inject.assistedinject.Assisted;
import com.google.inject.assistedinject.AssistedInject;
import lombok.Getter;

/**
 * Provides the animation of fade in or fade out for component.
 */
public class FadeAnimation implements ComponentAnimation {

    private final float duration;
    private final float startAlpha;
    private final float endAlpha;

    private float currentExecutionTime = 0;

    @Getter
    private float alpha;

    @AssistedInject
    public FadeAnimation(@Assisted("startAlpha") float startAlpha,
                         @Assisted("endAlpha") float endAlpha,
                         @Assisted("duration") float duration) {
        this.duration = duration;
        this.startAlpha = startAlpha;
        this.endAlpha = endAlpha;
        alpha = startAlpha;
    }

    @Override
    public void act(float delta) {
        if (isCompleted()) {
            return;
        }

        currentExecutionTime += delta;

        if (this.currentExecutionTime >= duration) {
            alpha = endAlpha;
        } else {
            // calculates the percentage of current time
            float currentPercentage = this.currentExecutionTime / duration * 100;
            alpha = startAlpha + (endAlpha - startAlpha) / 100 * currentPercentage;
        }
    }

    @Override
    public void reset() {
        alpha = startAlpha;
        currentExecutionTime = 0;
    }

    @Override
    public void destroy() {
    }

    private boolean isCompleted() {
        return this.currentExecutionTime >= duration;
    }
}

File: core/src/com/alta_v2/rendering/common/component/animation/FadeAnimation.kt
package com.alta_v2.rendering.common.component.animation

import com.google.inject.assistedinject.Assisted
import com.google.inject.assistedinject.AssistedInject

/**
 * Provides the animation of fade in or fade out for component.
 */
class FadeAnimation @AssistedInject constructor(@param:Assisted("startAlpha") private val startAlpha: Float,
                                                @param:Assisted("endAlpha") private val endAlpha: Float,
                                                @param:Assisted("duration") private val duration: Float) : ComponentAnimation {
    private var currentExecutionTime = 0f

    private var alpha: Float

    init {
        alpha = startAlpha
    }

    fun getAlpha() = alpha

    override fun act(delta: Float) {
        if (isCompleted) {
            return
        }
        currentExecutionTime += delta
        alpha = if (currentExecutionTime >= duration) {
            endAlpha
        } else {
            // calculates the percentage of current time
            val currentPercentage = currentExecutionTime / duration * 100
            startAlpha + (endAlpha - startAlpha) / 100 * currentPercentage
        }
    }

    override fun reset() {
        alpha = startAlpha
        currentExecutionTime = 0f
    }

    override fun destroy() {}

    private val isCompleted: Boolean
        get() = currentExecutionTime >= duration
}